Backend/lib/location-updates/src/Lib/LocationUpdates/Internal.hs

```diff
@@ -129,31 +139,42 @@ recalcDistanceBatches h@RideInterpolationHandler {..} ending driverId estDist pi
-      else updateDistance driverId (metersToHighPrecMeters estDist) 0 0
+  let snapToRoadCallCondition = routeDeviation || pickupDropOutsideThreshold
+  if ending
+    then do
+      if snapToRoadCallCondition
+        then do
+          currSnapToRoadState <- processSnapToRoadCall
+          updateDistance driverId currSnapToRoadState.distanceTravelled currSnapToRoadState.googleSnapToRoadCalls currSnapToRoadState.osrmSnapToRoadCalls
+        else updateDistance driverId (metersToHighPrecMeters estDist) 0 0
+    else do
+      isAtLeastBatchPlusOne <- atLeastBatchPlusOne
+      when (snapToRoadCallCondition && isAtLeastBatchPlusOne) $ do
+        currSnapToRoadState <- processSnapToRoadCall
+        Redis.setExp (onRideSnapToRoadStateKey driverId) currSnapToRoadState 21600 -- 6 hours
   where
-    -- atLeastBatchPlusOne = (> batchSize) <$> getWaypointsNumber driverId
     pointsRemaining = (> 0) <$> getWaypointsNumber driverId
-    continueCondition = if ending then pointsRemaining else return False
+    atLeastBatchPlusOne = (> batchSize) <$> getWaypointsNumber driverId
 
-    recalcDistanceBatches' acc googleSnapToRoadCalls osrmSnapToRoadCalls snapToRoadCallFailed = do
-      batchLeft <- continueCondition
-      if batchLeft
+    recalcDistanceBatches' snapToRoad'@SnapToRoadState {..} snapToRoadCallFailed isBatchLeft = do
+      if isBatchLeft
         then do
           (dist, servicesUsed, snapCallFailed) <- recalcDistanceBatchStep h driverId
           let googleCalled = Google `elem` servicesUsed
               osrmCalled = OSRM `elem` servicesUsed
           if snapCallFailed
-            then pure (acc, googleSnapToRoadCalls + fromBool googleCalled, osrmSnapToRoadCalls + fromBool osrmCalled, snapCallFailed)
+            then pure (SnapToRoadState distanceTravelled (googleSnapToRoadCalls + fromBool googleCalled) (osrmSnapToRoadCalls + fromBool osrmCalled), snapCallFailed)
             else do
-              recalcDistanceBatches' (acc + dist) (googleSnapToRoadCalls + fromBool googleCalled) (osrmSnapToRoadCalls + fromBool osrmCalled) snapCallFailed
-        else pure (acc, googleSnapToRoadCalls, osrmSnapToRoadCalls, snapToRoadCallFailed)
+              batchLeft <- atLeastBatchPlusOne
+              recalcDistanceBatches' (SnapToRoadState (distanceTravelled + dist) (googleSnapToRoadCalls + fromBool googleCalled) (osrmSnapToRoadCalls + fromBool osrmCalled)) snapCallFailed batchLeft
+        else pure (snapToRoad', snapToRoadCallFailed)
+
+    processSnapToRoadCall = do
+      prevSnapToRoadState :: SnapToRoadState <- (Redis.safeGet $ onRideSnapToRoadStateKey driverId) >>= pure . fromMaybe (SnapToRoadState 0 0 0)
+      arePointsRemaining <- pointsRemaining
+      (currSnapToRoadState, snapToRoadCallFailed) <- recalcDistanceBatches' prevSnapToRoadState False arePointsRemaining
+      when snapToRoadCallFailed $ throwError $ InternalError $ "Snap to road call failed for driverId =" <> driverId.getId
```

Comment:
We should not throw error, if snap to road is failing, right? End ride will also fail in that case. It should be handled properly as before.

Reply:
@hkmangla we throw error here and in parent function we have a catchAll that sets isDistanceCalculation flag to true. So that henceforth, we dont enter this flow. We used to throw this error earlier too I just moved the placement.